File: program/src/state/cash.rs
use borsh::{BorshDeserialize, BorshSerialize};
use solana_program::{
    borsh1::try_from_slice_unchecked,
    msg,
    program_error::ProgramError,
    program_pack::{IsInitialized, Pack, Sealed},
    pubkey::Pubkey,
};

use crate::error::CashError;

use super::AccountType;

pub const CASH_DATA_SIZE: usize = 195;

#[repr(C)]
#[derive(BorshSerialize, BorshDeserialize, Debug, PartialEq, Clone, Default)]
#[borsh(use_discriminant=true)]
pub enum CashState {
    #[default]
    Initialized = 0,
    Redeemed,
    Redeeming,
    Canceled,
}

#[repr(C)]
#[derive(BorshSerialize, BorshDeserialize, Debug, PartialEq, Clone, Default)]
#[borsh(use_discriminant=true)]
pub enum DistributionType {
    #[default]
    Fixed = 0,
    Random,
    Weighted,
    Equal,
}

#[repr(C)]
#[derive(Debug, Clone, PartialEq, BorshSerialize, BorshDeserialize, Default)]
pub struct Cash {
    pub account_type: AccountType,
    pub authority: Pubkey,
    pub state: CashState,
    pub amount: u64,
    pub fee_bps: u16,
    pub network_fee: u64,
    pub base_fee_to_redeem: u64,
    pub rent_fee_to_redeem: u64,
    pub remaining_amount: u64,
    pub distribution_type: DistributionType,//77
    pub owner: Pubkey,
    pub mint: Pubkey,
    pub total_redemptions: u16,
    pub max_num_redemptions: u16,
    pub min_amount: u64,
    pub fingerprint_enabled: bool,
    pub pass_key: Option<Pubkey>,//187
    pub total_weight_ppm: u32,
}

impl Cash {
    pub const PREFIX: &'static str = "cash";
    pub fn redeemed(&self) -> bool {
        self.state == CashState::Redeemed
    }
    pub fn redeeming(&self) -> bool {
        self.state == CashState::Redeeming
    }
    pub fn canceled(&self) -> bool {
        self.state == CashState::Canceled
    }
    pub fn initialized(&self) -> bool {
        self.state == CashState::Initialized
    }
    pub fn is_fully_redeemed(&self) -> Result<bool, CashError> {
        Ok(self.total_redemptions == self.max_num_redemptions
            || self.remaining_amount == 0
            || self.remaining_amount < self.min_total_required()?)
    }
    pub fn max_fee_to_redeem(&self) -> Result<u64, CashError> {
        self.base_fee_to_redeem.checked_add(self.rent_fee_to_redeem).ok_or(CashError::Overflow)
    }
    pub fn max_num_redemptions_remaining(&self) -> Result<u16, CashError> {
        self.max_num_redemptions
            .checked_sub(self.total_redemptions)
            .ok_or(CashError::Overflow)
    }

    pub fn min_total_required(&self) -> Result<u64, CashError> {
        Ok(self.min_amount * self.max_num_redemptions_remaining()? as u64)
    }
}

impl IsInitialized for Cash {
    fn is_initialized(&self) -> bool {
        self.initialized() || self.redeeming() || self.redeemed() || self.canceled()
    }
}

impl Sealed for Cash {}

impl Pack for Cash {
    const LEN: usize = CASH_DATA_SIZE;

    fn pack_into_slice(&self, dst: &mut [u8]) {
        let mut slice = dst;
        self.serialize(&mut slice).unwrap()
    }

    fn unpack_from_slice(src: &[u8]) -> Result<Self, ProgramError> {
        if src.len() != Self::LEN {
            msg!("Failed to deserialize");
            return Err(ProgramError::InvalidAccountData);
        }

        let result: Self = try_from_slice_unchecked(src)?;

        Ok(result)
    }
}

File: src/accounts/cash.ts
import {
  Borsh,
  AnyPublicKey,
  ERROR_INVALID_OWNER,
  Account,
  StringPublicKey,
} from '@metaplex-foundation/mpl-core';
import { AccountInfo, Commitment, Connection, PublicKey } from '@solana/web3.js';
import BN from 'bn.js';
import bs58 from 'bs58';
import { CashProgram } from '../cash_program';
import { AccountType } from './account';

export const MAX_CASH_LINK_DATA_LEN = 194;

export enum CashState {
  Initialized = 0,
  Redeemed = 1,
  Redeeming = 2,
  Canceled = 3,
}

export enum CashDistributionType {
  Fixed = 0,
  Random = 1,
  Weighted = 2,
  Equal = 3,
}

export type CashDataArgs = {
  accountType: AccountType;
  authority: StringPublicKey;
  state: CashState;
  amount: BN;
  feeBps: number;
  networkFee: BN;
  baseFeeToRedeem: BN;
  rentFeeToRedeem: BN;
  remainingAmount: BN;
  distributionType: CashDistributionType;
  owner: number;
  mint: StringPublicKey;
  totalRedemptions: BN;
  maxNumRedemptions: BN;
  minAmount: BN;
  fingerprintEnabled: boolean;
  passKey?: StringPublicKey;
  totalWeightPpm: number;
};

export class CashData extends Borsh.Data<CashDataArgs> {
  static readonly SCHEMA = CashData.struct([
    ['accountType', 'u8'],
    ['authority', 'pubkeyAsString'],
    ['state', 'u8'],
    ['amount', 'u64'],
    ['feeBps', 'u16'],
    ['networkFee', 'u64'],
    ['baseFeeToRedeem', 'u64'],
    ['rentFeeToRedeem', 'u64'],
    ['remainingAmount', 'u64'],
    ['distributionType', 'u8'],
    ['owner', 'pubkeyAsString'],
    ['mint', 'pubkeyAsString'],
    ['totalRedemptions', 'u16'],
    ['maxNumRedemptions', 'u16'],
    ['minAmount', 'u64'],
    ['fingerprintEnabled', 'u8'],
    ['passKey', { kind: 'option', type: 'pubkeyAsString' }],
    ['totalWeightPpm', 'u32'],
  ]);
  accountType: AccountType;
  authority: StringPublicKey;
  state: CashState;
  amount: BN;
  feeBps: number;
  networkFee: BN;
  baseFeeToRedeem: BN;
  rentFeeToRedeem: BN;
  remainingAmount: BN;
  distributionType: CashDistributionType;
  owner: StringPublicKey;
  mint: StringPublicKey;
  totalRedemptions: number;
  maxNumRedemptions: number;
  minAmount: BN;
  fingerprintEnabled: boolean;
  passKey?: StringPublicKey;
  totalWeightPpm: number;

  constructor(args: CashDataArgs) {
    super(args);
  }
}

export class Cash extends Account<CashData> {
  static readonly PREFIX = 'cash';
  constructor(pubkey: AnyPublicKey, info: AccountInfo<Buffer>) {
    super(pubkey, info);
    this.data = CashData.deserialize(this.info.data);
    if (!this.assertOwner(CashProgram.PUBKEY)) {
      throw ERROR_INVALID_OWNER();
    }
  }

  static getPDA(reference: string) {
    const [pubKey] = CashProgram.cashAccount(reference);
    return pubKey;
  }

  static async findMany(
    connection: Connection,
    filters: {
      authority?: AnyPublicKey;
      state?: CashState;
    } = {},
    commitment?: Commitment,
  ) {
    const baseFilters = [
      // Filter for Cash by account type
      {
        memcmp: {
          offset: 0,
          bytes: bs58.encode(Buffer.from([AccountType.Cash])),
        },
      },
      // Filter for assigned to authority
      filters.authority && {
        memcmp: {
          offset: 1,
          bytes: new PublicKey(filters.authority).toBase58(),
        },
      },
      // Filter by state
      filters.state && {
        memcmp: {
          offset: 33,
          bytes: bs58.encode(Buffer.from([filters.state])),
        },
      },
    ].filter(Boolean);

    return (
      await CashProgram.getProgramAccounts(connection, { filters: baseFilters, commitment })
    ).map((account) => Cash.from(account));
  }
}
